com.hazelcast.jet.elastic.LocalElasticSinkTest.given_singleDocument_whenWriteToElasticSink_then_singleDocumentInIndex

_master_ (commit a1740b07b1104d537b72be24b40230d1bc01ca36)

Failed on Sonar nightly build: http://jenkins.hazelcast.com/job/jet-oss-master-sonar-nightly/581/testReport/junit/com.hazelcast.jet.elastic/LocalElasticSinkTest/given_singleDocument_whenWriteToElasticSink_then_singleDocumentInIndex/

Stacktrace:
```
java.lang.OutOfMemoryError: GC overhead limit exceeded
```

Standard output:
```
2020-08-12 15:17:30,268 [ WARN] [elasticsearch[_client_][transport_client_boss][T#40]] [i.n.u.c.DefaultPromise]: An exception was thrown by org.elasticsearch.transport.netty4.Netty4Transport$$Lambda$824/1490715251.operationComplete()
java.lang.OutOfMemoryError: GC overhead limit exceeded
2020-08-12 15:21:48,880 [ERROR] [hz.hungry_khayyam.partition-operation.thread-63] [c.h.s.i.o.i.OperationExecutorImpl]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Java heap space
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:18,463 [ WARN] [hz.hungry_khayyam.partition-operation.thread-63] [c.h.s.i.o.i.OperationExecutorImpl]: Logging took 503211 ms.
2020-08-12 15:24:21,105 [ WARN] [main] [c.h.i.i.Node]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Terminating forcefully...
2020-08-12 15:24:21,105 [ INFO] [main] [c.h.i.i.Node]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Shutting down connection manager...
2020-08-12 15:24:21,118 [ INFO] [main] [c.h.i.i.Node]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Shutting down node engine...
2020-08-12 15:24:21,512 [ INFO] [hz.hungry_khayyam.HealthMonitor] [c.h.i.d.HealthMonitor]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] processors=72, physical.memory.total=377.6G, physical.memory.free=119.4G, swap.space.total=4.0G, swap.space.free=3.9G, heap.memory.used=797.0M, heap.memory.free=113.7M, heap.memory.total=910.5M, heap.memory.max=910.5M, heap.memory.used/total=87.47%, heap.memory.used/max=87.47%, minor.gc.count=91, minor.gc.time=1495ms, major.gc.count=21470, major.gc.time=4095502ms, load.process=11.63%, load.system=21.91%, load.systemAverage=21.23, thread.count=4712, thread.peakCount=5146, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=27, executor.q.priorityOperation.size=0, operations.completed.count=545, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=7, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
2020-08-12 15:18:16,713 [ERROR] [hz.hungry_khayyam.partition-operation.thread-19] [c.h.s.i.o.i.OperationExecutorImpl]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Failed to process: com.hazelcast.spi.impl.operationexecutor.impl.TaskBatch@6bf2ec92 on: hz.hungry_khayyam.partition-operation.thread-19
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:21,515 [ WARN] [hz.hungry_khayyam.partition-operation.thread-19] [c.h.s.i.o.i.OperationExecutorImpl]: Logging took 364801 ms.
2020-08-12 15:24:21,539 [ WARN] [hz.hungry_khayyam.InvocationMonitorThread] [c.h.s.i.o.i.InvocationMonitor]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] MonitorInvocationsTask delayed 24104 ms
2020-08-12 15:24:21,539 [ WARN] [elasticsearch[_client_][transport_client_boss][T#10]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:21,539 [ WARN] [elasticsearch[_client_][transport_client_boss][T#5]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:21,539 [ WARN] [elasticsearch[_client_][transport_client_boss][T#12]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:21,539 [ WARN] [elasticsearch[_client_][transport_client_boss][T#13]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: GC overhead limit exceeded
2020-08-12 15:24:21,539 [ WARN] [elasticsearch[_client_][transport_client_boss][T#9]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: GC overhead limit exceeded
2020-08-12 15:24:21,539 [ WARN] [hz.hungry_khayyam.InvocationMonitorThread] [c.h.s.i.o.i.InvocationMonitor]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] MonitorInvocationsTask delayed 4258546 ms
2020-08-12 15:24:21,540 [ WARN] [hz.hungry_khayyam.InvocationMonitorThread] [c.h.s.i.o.i.InvocationMonitor]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] BroadcastOperationControlTask delayed 4274775 ms
2020-08-12 15:24:21,540 [ WARN] [elasticsearch[_client_][transport_client_boss][T#4]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:21,540 [ WARN] [elasticsearch[_client_][transport_client_boss][T#7]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: GC overhead limit exceeded
2020-08-12 15:24:21,540 [ WARN] [elasticsearch[_client_][transport_client_boss][T#4]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: Java heap space
2020-08-12 15:24:21,540 [ WARN] [elasticsearch[_client_][transport_client_boss][T#4]] [i.n.u.ReferenceCountUtil]: Failed to release a message: PooledDirectByteBuf(freed)
java.lang.OutOfMemoryError: GC overhead limit exceeded
2020-08-12 15:24:21,540 [ WARN] [elasticsearch[_client_][transport_client_boss][T#17]] [i.n.c.AbstractChannelHandlerContext]: Failed to mark a promise as failure because it has failed already: DefaultChannelPromise@5fd6e7bd(failure: java.lang.OutOfMemoryError: Java heap space), unnotified cause: java.lang.OutOfMemoryError: Java heap space

java.lang.OutOfMemoryError: GC overhead limit exceeded
2020-08-12 15:24:21,539 [ERROR] [elasticsearch[_client_][transport_client_boss][T#13]] [o.e.t.n.Netty4Utils]: fatal error on the network layer
	at org.elasticsearch.transport.netty4.Netty4Utils.maybeDie(Netty4Utils.java:184)
	at org.elasticsearch.transport.netty4.Netty4Transport.lambda$sendMessage$4(Netty4Transport.java:406)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:481)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:122)
	at io.netty.util.internal.PromiseNotificationUtil.tryFailure(PromiseNotificationUtil.java:64)
	at io.netty.channel.AbstractChannelHandlerContext.notifyOutboundHandlerException(AbstractChannelHandlerContext.java:837)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:740)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.access$1900(AbstractChannelHandlerContext.java:38)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.write(AbstractChannelHandlerContext.java:1089)
	at io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.write(AbstractChannelHandlerContext.java:1136)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:1078)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
2020-08-12 15:24:21,539 [ERROR] [elasticsearch[_client_][transport_client_boss][T#9]] [o.e.t.n.Netty4Utils]: fatal error on the network layer
	at org.elasticsearch.transport.netty4.Netty4Utils.maybeDie(Netty4Utils.java:184)
	at org.elasticsearch.transport.netty4.Netty4Transport.lambda$sendMessage$4(Netty4Transport.java:406)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:481)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:122)
	at io.netty.util.internal.PromiseNotificationUtil.tryFailure(PromiseNotificationUtil.java:64)
	at io.netty.channel.AbstractChannelHandlerContext.notifyOutboundHandlerException(AbstractChannelHandlerContext.java:837)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:740)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.access$1900(AbstractChannelHandlerContext.java:38)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.write(AbstractChannelHandlerContext.java:1089)
	at io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.write(AbstractChannelHandlerContext.java:1136)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:1078)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/extensions/elasticsearch/elasticsearch-5/src/test/java/com/hazelcast/jet/elastic/BaseElasticTest.java
+++ b/extensions/elasticsearch/elasticsearch-5/src/test/java/com/hazelcast/jet/elastic/BaseElasticTest.java
@@ -36,6 +36,7 @@ import org.elasticsearch.action.index.IndexRequest;
 import org.elasticsearch.action.search.SearchRequest;
 import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.action.support.WriteRequest.RefreshPolicy;
+import org.elasticsearch.client.RestClient;
 import org.elasticsearch.client.RestClientBuilder;
 import org.elasticsearch.client.RestHighLevelClient;
 import org.elasticsearch.client.transport.TransportClient;
@@ -43,6 +44,7 @@ import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.transport.InetSocketTransportAddress;
 import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.transport.client.PreBuiltTransportClient;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -73,6 +75,7 @@ public abstract class BaseElasticTest {
 
     protected static final int BATCH_SIZE = 42;
 
+    protected RestClient lowLevelClient;
     protected RestHighLevelClient elasticClient;
     protected TransportClient transportClient;
 
@@ -81,8 +84,14 @@ public abstract class BaseElasticTest {
 
     @Before
     public void setUpBase() {
+        // The transport client is not friendly to creating a lot of instances and closing them due to netty pooling
+        // See https://github.com/elastic/elasticsearch/issues/26048
+        System.setProperty("io.netty.allocator.type", "unpooled");
+
         if (elasticClient == null) {
-            elasticClient = new RestHighLevelClient(elasticClientSupplier().get().build());
+            // Need the lowLevelClient instance to close it at the end
+            lowLevelClient = elasticClientSupplier().get().build();
+            elasticClient = new RestHighLevelClient(lowLevelClient);
             transportClient = new PreBuiltTransportClient(Settings.EMPTY)
                     .addTransportAddress(new InetSocketTransportAddress(ElasticSupport.elastic.get().getTcpHost()));
 
@@ -96,6 +105,26 @@ public abstract class BaseElasticTest {
         results.clear();
     }
 
+    @After
+    public void tearDown() throws Exception {
+        // We can't close the RestHighLevelClient as it doesn't have any close()/destroy() method
+        // Close the underlying lowLevelClient instead
+        if (lowLevelClient != null) {
+            try {
+                lowLevelClient.close();
+            } finally {
+                lowLevelClient = null;
+            }
+        }
+        if (transportClient != null) {
+            try {
+                transportClient.close();
+            } finally {
+                transportClient = null;
+            }
+        }
+    }
+
     /**
      * RestHighLevelClient supplier, it is used to
      * - create a client before each test for use by all methods from this class interacting with elastic
@@ -103,7 +132,7 @@ public abstract class BaseElasticTest {
      */
     protected SupplierEx<RestClientBuilder> elasticClientSupplier() {
         return ElasticSupport.elasticClientSupplier();
-    };
+    }
 
     protected abstract JetInstance createJetInstance();
 

--- a/extensions/elasticsearch/elasticsearch-5/src/test/java/com/hazelcast/jet/elastic/LocalElasticSinkTest.java
+++ b/extensions/elasticsearch/elasticsearch-5/src/test/java/com/hazelcast/jet/elastic/LocalElasticSinkTest.java
@@ -30,6 +30,7 @@ import org.elasticsearch.action.support.WriteRequest.RefreshPolicy;
 import org.elasticsearch.client.RestClient;
 import org.elasticsearch.client.RestClientBuilder;
 import org.junit.After;
+import org.junit.AfterClass;
 import org.junit.Test;
 
 import java.io.IOException;
@@ -89,4 +90,8 @@ public class LocalElasticSinkTest extends CommonElasticSinksTest {
         }
     }
 
+    @AfterClass
+    public static void afterClass() throws Exception {
+        ClientHolder.elasticClients.clear();
+    }
 }

--- a/extensions/elasticsearch/elasticsearch-6/src/test/java/com/hazelcast/jet/elastic/BaseElasticTest.java
+++ b/extensions/elasticsearch/elasticsearch-6/src/test/java/com/hazelcast/jet/elastic/BaseElasticTest.java
@@ -40,6 +40,7 @@ import org.elasticsearch.client.RestClientBuilder;
 import org.elasticsearch.client.RestHighLevelClient;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.search.SearchHit;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -87,6 +88,17 @@ public abstract class BaseElasticTest {
         }
         results = jet.getList("results");
         results.clear();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        if (elasticClient != null) {
+            try {
+                elasticClient.close();
+            } finally {
+                elasticClient = null;
+            }
+        }
     }
 
     /**

--- a/extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/BaseElasticTest.java
+++ b/extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/BaseElasticTest.java
@@ -40,6 +40,7 @@ import org.elasticsearch.client.indices.CreateIndexRequest;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.reindex.DeleteByQueryRequest;
 import org.elasticsearch.search.SearchHit;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -88,6 +89,17 @@ public abstract class BaseElasticTest {
         }
         results = jet.getList("results");
         results.clear();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        if (elasticClient != null) {
+            try {
+                elasticClient.close();
+            } finally {
+                elasticClient = null;
+            }
+        }
     }
 
     /**

